### PR TITLE
SMP protocol commands encoding/decoding

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: pubspec.lock [hunk 1]
```diff
@@ -1,6 +1,27 @@
 # Generated by pub
```
Comment (alextekartik) [inline pubspec.lock:1]: I would not put this file in source control (you could keep it only for the main application but I would say not for libraries)

File: packages/simplexmq/lib/simplexmq.dart [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/// Support for doing something awesome.
+///
+/// More dartdocs go here.
+library simplexmq;
+
+export "src/protocol.dart";
```
Comment (alextekartik) [inline packages/simplexmq/lib/simplexmq.dart:6]: You might not want to export all symbols (such as messageP, smpCommandP...), you can add show/hide to selectively show only some symbols or hide specific ones.

**Review by alextekartik — COMMENTED** [review]
Looks all good and tested!

**Review by efim-poberezkin — APPROVED** [review]

File: pubspec.yaml [hunk 3]
```diff
@@ -18,7 +18,7 @@ publish_to: 'none' # Remove this line if you wish to publish to pub.dev
 version: 1.0.0+1
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply pubspec.yaml:18]: would it be in one of the packages then?

File: packages/simplexmq/analysis_options.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# 'package:lints/core.yaml'. These are just the most critical lints
+# (the recommended set includes the core lints).
+# The core lints are also what is used by pub.dev for scoring packages.
+
+include: package:lints/recommended.yaml
+
+linter:
+  rules:
+    prefer_double_quotes: true
+    constant_identifier_names: false
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply packages/simplexmq/analysis_options.yaml:19]: cool, it was a bit annoying not allowing names char_space but I guess I'll get used to it :)

Is there a way to have shared lint settings across all projects? It doesn't seem to inherit settings from the root.

File: pubspec.lock (hunk 1 above)
Comment (epoberezkin) [reply pubspec.lock:1]: it's for the main app I think, I removed from the library.

File: pubspec.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alextekartik) [reply pubspec.yaml:18]: Yes the flutter app could be in `packages`.

File: packages/simplexmq/lib/simplexmq.dart (hunk 2 above)
Comment (epoberezkin) [reply packages/simplexmq/lib/simplexmq.dart:6]: I didn't figure out this syntax yet - can they be shown/hidden at the place of export? I thought it's at the place of import when I read the doc...

Also, is there a way to import the whole namespace as a name, to reference the symbols as Namespace.name?

File: packages/simplexmq/analysis_options.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (alextekartik) [reply packages/simplexmq/analysis_options.yaml:19]: In theory, it should inherit.  But you can also define lints in a core package (let's name it `simplex_core` for example putting your shared rules  in `lib/lints/recommended.yaml`) and reference it from other package:

```yaml
include: package:simplex_core/lints/recommended.yaml
```

File: packages/simplexmq/lib/simplexmq.dart (hunk 2 above)
Comment (alextekartik) [reply packages/simplexmq/lib/simplexmq.dart:6]: ExamYou can control in both import and export.

In export, to hide some sympols.
```yaml
// Export and hide some symbols
export "src/protocol.dart" hide smpCommandP, messageP;
```
Examples for import
```yaml
// Import and hide some symbols
import "package:simplexmq/simplexmq.dart" hide smpCommandP, messageP;

// Import only some symbols
import "package:simplexmq/simplexmq.dart" hide CmdErrorType;

// Import in a namespace
import "package:simplexmq/simplexmq.dart" as simplexmq;
```

File: pubspec.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (epoberezkin) [reply pubspec.yaml:18]: ok, will move it later

File: packages/simplexmq/analysis_options.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (epoberezkin) [reply packages/simplexmq/analysis_options.yaml]: Moving analysis_options to packages root worked, the rules outside of packages were ignored though... strange
Added all these rules, kept only constant_identifier_names disabled to allow uppercase names in enums.

File: packages/simplexmq/lib/simplexmq.dart (hunk 2 above)
Comment (epoberezkin) [reply packages/simplexmq/lib/simplexmq.dart:6]: Cool.
Is there a way to list exported symbols inside protocol.dart itself?

**Pull request merged**